ts/private/ts_project.bzl

```diff
@@ -225,23 +225,53 @@ This is an error because Bazel does not run actions unless their outputs are nee
     else:
         # We must avoid tsc writing any JS files in this case, as tsc was only run for typings, and some other
         # action will try to write the JS files. We must avoid collisions where two actions write the same file.
-        arguments.add("--emitDeclarationOnly")
+        if not ctx.attr.no_emit:
+            arguments.add("--emitDeclarationOnly")
 
         # We don't produce any DefaultInfo outputs in this case, because we avoid running the tsc action
         # unless the output_types are requested.
         default_outputs = []
 
     srcs_tsconfig_deps = ctx.attr.srcs + [ctx.attr.tsconfig] + ctx.attr.deps
 
-    inputs_depset = depset()
-    if len(outputs) > 0:
-        transitive_inputs.append(_gather_types_from_js_infos(srcs_tsconfig_deps))
+    transitive_inputs.append(_gather_types_from_js_infos(srcs_tsconfig_deps))
 
-        inputs_depset = depset(
-            copy_files_to_bin_actions(ctx, inputs),
-            transitive = transitive_inputs,
-        )
+    inputs_depset = depset(
+        copy_files_to_bin_actions(ctx, inputs),
+        transitive = transitive_inputs,
+    )
+
+    if ctx.attr.no_emit:
+        verb = "Type-checking"
+
+        # Validation actions still need to produce some output, so we output a .validation file
+        # that ends up in the _validation output group.
```

Comment:
add to the comment that the content of that file will be any stdout written by `tsc`